pymacaroons/serializers/binary_serializer.py

```diff
@@ -113,3 +237,85 @@ def _depacketize(self, packet):
         key = packet.split(b' ')[0]
         value = packet[len(key) + 1:-1]
         return (key, value)
+
+
+def _parse_section_v2(data):
+    ''' Parses a sequence of packets in data.
+
+    The sequence is terminated by a packet with a field type of EOS
+    :param data bytes to be deserialized.
+    :return: the rest of data and an array of packet V2
+    '''
+
+    from pymacaroons.exceptions import MacaroonDeserializationException
+
+    prev_field_type = -1
+    packets = []
+    while True:
+        if len(data) == 0:
+            raise MacaroonDeserializationException(
+                'section extends past end of buffer')
+        rest, field_type, payload = _parse_packet_v2(data)
+        if field_type == EOS:
+            return rest, packets
+        if field_type <= prev_field_type:
+            raise MacaroonDeserializationException('fields out of order')
+        packets.append((field_type, payload))
+        prev_field_type = field_type
+        data = rest
+
+
+def _parse_packet_v2(data):
+    ''' Parses a V2 data package at the start of the given data.
+
+    The format of a packet is as follows:
+
+    field_type(varint) payload_len(varint) data[payload_len bytes]
+
+    apart from EOS which has no payload_en or data (it's a single zero byte).
+    '''
+    from pymacaroons.exceptions import MacaroonDeserializationException
+
+    ft, n = _decode_uvarint(data)
+    data = data[n:]
+    if ft == EOS:
+        return data, ft, None
+    payload_len, n = _decode_uvarint(data)
+    data = data[n:]
+    if payload_len > len(data):
+        raise MacaroonDeserializationException(
+            'field data extends past end of buffer')
+    return data[payload_len:], ft, data[0:payload_len]
+
+
+def _encode_uvarint(n, data):
+    ''' Encodes integer into variable-length format into data.'''
+    if n < 0:
+        raise ValueError('only support positive integer')
+    while True:
+        this_byte = n & 127
+        n >>= 7
+        if n == 0:
+            data.append(this_byte)
+            break
+        data.append(this_byte | 128)
+
+
+def _decode_uvarint(data):
+    ''' Decode a variable -length integer.
+
+    Reads a sequence of unsigned integer byte and decodes them into an integer
+    in variable-length format and returns it and the length read.
+    '''
+    n = 0
+    shift = 0
+    length = 0
+    for b in data:
+        if not isinstance(b, int):
```

Comment:
not AFAICS:

	% python
	Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
	[GCC 5.4.0 20160609] on linux2
	Type "help", "copyright", "credits" or "license" for more information.
	>>> b=bytearray('hello', 'utf8'
	... )
	>>> 
	>>> b[0]
	104
	>>> type(b[0])
	<type 'int'>